### PR TITLE
JER integration into conditions database

### PR DESCRIPTION
**Pull request by blinkseb**
This PR adds the necessary machinery to store Jet energy Resolutions and associated Scale Factors to the conditions database.

Values are parsed from text files using the same format as Jet Energy Corrections text files. Two new records are added, `JetResolutionRcd` and `JetResolutionScaleFactorRcd`, storing respectively resolutions and scale factors. Both resolutions and scale factors are converted to a common object, `JetResolutionObject`, which is serialized into the corresponding records.

Facility are provided to create a database from text files [1] and to create text files from a database or a GT [2].

Users can access values using the classes `JetResolution` and `JetResolutionScaleFactors`. A static function is provided to hide the boilerplate to retrieve the object from the condition database. A demo analyzer is included as an how-to tutorial [3]

This interface has been presented and validated yesterday to the JERC meeting.

I targeted CMSSW 7.5.x during my development, bu we'd like to see this also back-ported to 7.4.x

[1] https://github.com/blinkseb/cmssw/blob/665bd5d702f5ed976a898b7f8c426a5955eaf103/JetMETCorrections/Modules/test/JetResolutionDBWriter_cfg.py
[2] https://github.com/blinkseb/cmssw/blob/665bd5d702f5ed976a898b7f8c426a5955eaf103/JetMETCorrections/Modules/test/JetResolutionLocalDBReader_cfg.py
[3] https://github.com/blinkseb/cmssw/blob/665bd5d702f5ed976a898b7f8c426a5955eaf103/JetMETCorrections/Modules/plugins/JetResolutionDemo.cc

### CODE REVIEW

File: CondFormats/JetMETObjects/src/JetResolutionObject.cc [hunk 1]
```diff
@@ -0,0 +1,437 @@
+#include <CondFormats/JetMETObjects/interface/JetResolutionObject.h>
+#include <CondFormats/JetMETObjects/interface/Utilities.h>
+#include <FWCore/Framework/interface/EventSetup.h>
+#include <FWCore/Framework/interface/ESHandle.h>
+#include <FWCore/Utilities/interface/EDMException.h>
+
+#include <CondFormats/DataRecord/interface/JetResolutionRcd.h>
+#include <CondFormats/DataRecord/interface/JetResolutionScaleFactorRcd.h>
+
```
Comment (diguida) [inline CondFormats/JetMETObjects/src/JetResolutionObject.cc]: unneeded includes: here you do not deliver the object via `EventSetup`, you are implementing the interface of the entity classes.

File: CondFormats/JetMETObjects/BuildFile.xml [hunk 2]
```diff
@@ -2,6 +2,7 @@
 <use   name="FWCore/ParameterSet"/>
 <use   name="FWCore/Utilities"/>
 <use   name="CondFormats/Serialization"/>
+<use   name="CondFormats/DataRecord"/>
```
Comment (diguida) [inline CondFormats/JetMETObjects/BuildFile.xml]: spurious dependency: if you remove the includes as in https://github.com/cms-sw/cmssw/pull/10350/files#r35481619 this is not needed anymore.

File: CondFormats/JetMETObjects/bin/BuildFile.xml [hunk 3]
```diff
@@ -3,3 +3,6 @@
 <use   name="root"/>
 <bin   name="JetResolution_t" file="JetResolution_t.cc">
 </bin>
+<bin   name="test_jet_resolution" file="test_jet_resolution.cc"/>
+<bin   name="test_jet_resolution_sf" file="test_jet_resolution_sf.cc"/>
+<bin   name="draw_jec_jer_factors" file="draw_factors.cc"/>
```
Comment (diguida) [inline CondFormats/JetMETObjects/bin/BuildFile.xml]: All these executable are actually unit tests for the object. I suggest to move them all to the test directory and set them to be unit tests for this module.

File: CondFormats/JetMETObjects/data/Summer15_V5_MC_JER_AK4PFchs.txt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{1    JetEta     1     JetPt     sqrt([0]^2/x^2+[1]^2/x+[2]^2)    Resolution} 
+-5.0   -3.2      5     10     200    4.5    0.1    0.05
+-3.2   -2.5      5     10     500    4.5    0.1    0.05
+-3.2   -1.3      5     10    1000    4.5    0.1    0.05
+-1.3    1.3      5     10    4000    4.5    0.1    0.05
+ 1.3    2.5      5     10    1000    4.5    0.1    0.05
+ 2.5    3.2      5     10     500    4.5    0.1    0.05
+ 3.2    5.0      5     10     200    4.5    0.1    0.05
```
Comment (diguida) [inline CondFormats/JetMETObjects/data/Summer15_V5_MC_JER_AK4PFchs.txt]: you should remove this from here and add it to the cms-data repo, and change the tests accordingly by using `edm::FileInPath`.

File: CondFormats/JetMETObjects/data/Summer15_V5_MC_JER_SF_AK4PFchs.txt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{1    JetAbsEta  0     None      Resolution} 
+0     0.5     3     1.079     1.053      1.105      
+0.5   1.1     3     1.099     1.071      1.127      
+1.1   1.7     3     1.121     1.092      1.150      
+1.7   2.3     3     1.208     1.162      1.254      
+2.3   2.8     3     1.254     1.192      1.316      
+2.8   3.2     3     1.395     1.332      1.458      
+3.2   5.0     3     1.056     0.865      1.247
```
Comment (diguida) [inline CondFormats/JetMETObjects/data/Summer15_V5_MC_JER_SF_AK4PFchs.txt]: you should remove this from here and add it to the cms-data repo, and change the tests accordingly by using `edm::FileInPath`.

File: JetMETCorrections/Modules/test/JetResolutionDemo_cfg.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import FWCore.ParameterSet.Config as cms
+
+process = cms.Process("myprocess")
+process.TFileService=cms.Service("TFileService", fileName=cms.string('JERplots.root'))
+
+##-------------------- Communicate with the DB -----------------------
+process.load('Configuration.StandardSequences.Services_cff')
+process.load("Configuration.StandardSequences.FrontierConditions_GlobalTag_condDBv2_cff")
```
Comment (diguida) [inline JetMETCorrections/Modules/test/JetResolutionDemo_cfg.py]: `process.load("Configuration.StandardSequences.FrontierConditions_GlobalTag_cff")`
I suggest this cfg to become an unit test.

File: JetMETCorrections/Modules/src/SealModule.cc [hunk 7]
```diff
@@ -58,6 +59,9 @@ DEFINE_FWK_EVENTSETUP_MODULE(JetCorrectionESChain);
 DEFINE_FWK_EVENTSETUP_MODULE(QGLikelihoodESProducer);
 DEFINE_FWK_EVENTSETUP_MODULE(QGLikelihoodSystematicsESProducer);
 
+DEFINE_FWK_EVENTSETUP_MODULE(JetResolutionESProducer);
+DEFINE_FWK_EVENTSETUP_MODULE(JetResolutionScaleFactorESProducer);
+
```
Comment (diguida) [inline JetMETCorrections/Modules/src/SealModule.cc]: Request for bug fix: please move:
https://github.com/blinkseb/cmssw/blob/jer_gt_rebased/JetMETCorrections/Modules/src/SealModule.cc#L32
https://github.com/blinkseb/cmssw/blob/jer_gt_rebased/JetMETCorrections/Modules/src/SealModule.cc#L33
(i.e. `DataProxy` registration in the `ESSource` for JEC) into `CondCore/JetMETPlugins/src/plugins.cc`

**Comment by diguida** [comment]
-1
Please fix the bugs reported in inline comments.
Please add macros for dropbox handling in `CondCore/Utilities`:
`FETCH_PAYLOAD_CASE` for each entity class in `CondCore/Utilities/src/CondDBFetch.cc`
`IMPORT_PAYLOAD_CASE` for each entity class in `CondCore/Utilities/src/CondDBImport.cc`

**Comment by diguida** [comment]
@blinkseb 75X cycle is closed for new developments. I suggest to do the following:
1. please fix the things we requested on this PR
2. then move the upgraded modules in a new PR 76X, and close this one;
3. once the 76X PR is approved, we can discuss of a 75X back port here.

File: CondFormats/JetMETObjects/src/JetResolutionObject.cc (hunk 1 above)
Comment (blinkseb) [inline CondFormats/JetMETObjects/src/JetResolutionObject.cc]: This is actually needed for the helper functions `JetResolution::get()` and `JetResolutionScaleFactor::get()`

File: CondFormats/JetMETObjects/bin/BuildFile.xml (hunk 3 above)
Comment (blinkseb) [inline CondFormats/JetMETObjects/bin/BuildFile.xml]: Sure, I'll do it

File: CondFormats/JetMETObjects/data/Summer15_V5_MC_JER_AK4PFchs.txt (hunk 4 above)
Comment (blinkseb) [inline CondFormats/JetMETObjects/data/Summer15_V5_MC_JER_AK4PFchs.txt]: Do you know who I should contact to create a repository on the cms-data org? I tried to find some documentation of the twiki but without any success...

Comment (mmusich) [inline CondFormats/JetMETObjects/data/Summer15_V5_MC_JER_AK4PFchs.txt]: the release manager (@degano?) should do it

**Comment by davidlange6** [comment]
given @diguida s comments, I'll close this

**Pull request closed without merging**